File: service/python/link-observer.py
from prometheus_client import start_http_server, Summary, Gauge
import random
import time
import requests

urls = ['https://httpstat.us/503','https://httpstat.us/200']

# Create a metric to track time spent and requests made.
MONITOR_PORT= 8001
REQUEST_SUMMARY = Summary('lo_external_url_response_ms', 'Time spent processing request',['url'])
RESPONSE_GAUGE = Gauge('lo_external_url_up', 'up = 1 all else 0', ['url'])

def process_request(url):
    with REQUEST_SUMMARY.labels(url).time():
        response = requests.get(url)
        return response.status_code

if __name__ == '__main__':
    # Start up the server to expose the metrics.
    start_http_server(MONITOR_PORT)
    # Generate some requests.
    while True:
        time.sleep(1)
        for url in urls:
            #RESPONSE_GAUGE_SET = RESPONSE_GAUGE.labels(url)
            resp = process_request(url)
            if resp == 200:
                RESPONSE_GAUGE.labels(url).set(1)
            else:
                RESPONSE_GAUGE.labels(url).set(0)

File: service/golang/link-observer.go
package main

import (
	"flag"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"log"
	"net/http"
	"time"
)

var (
	listeningAddress = flag.String("telemetry.address", "localhost:8003", "Address on which to expose metrics.")
	metricsEndpoint  = flag.String("telemetry.endpoint", "/metrics", "Path under which to expose metrics.")
	url              = [...]string{"https://httpstat.us/200", "https://httpstat.us/503"}
	REQUEST_SUMMARY  = prometheus.NewSummaryVec(prometheus.SummaryOpts{
		Name: "link_observer_url_response_ms",
		Help: "Agrigate Response time for URL probe",
	}, []string{"url"})
	RESPONSE_GAUGE = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "link_observer_url_up",
		Help: "Up == 1 (ret 200) Down == 0 (ret 503)",
	}, []string{"url"})
)

func process_request(url string) int {
	timer := prometheus.NewTimer(REQUEST_SUMMARY.WithLabelValues(url))
	resp, _ := http.Get(url)
	defer resp.Body.Close()
	defer timer.ObserveDuration()
	return resp.StatusCode
}

func probeSites() {
	var response_code int
	prometheus.Register(REQUEST_SUMMARY)
	prometheus.Register(RESPONSE_GAUGE)
	go func() {
		for true {
			for _, site := range url {
				time.Sleep(1)
				response_code = process_request(site)
				if response_code == 200 {
					RESPONSE_GAUGE.WithLabelValues(site).Set(1)
				} else {
					RESPONSE_GAUGE.WithLabelValues(site).Set(0)
				}

			}
		}
	}()
}

func main() {
	flag.Parse()
	probeSites()
	http.Handle(*metricsEndpoint, promhttp.Handler())
	log.Fatal(http.ListenAndServe(*listeningAddress, nil))

}
